File: solution/kotlin/week02-exercise2/src/main/kotlin/BookStore.kt
open class BookStore {
    private val inventory = BookInventory()

    fun addBook(title: String?, author: String?, copies: Int) {
        if (copies <= 0) return

        getBookBy(title, author)?.let {
            it.addCopies(copies)
        } ?: run {
            Book.tryCreateBook(title, author, copies)?.let {
                inventory.add(it)
            }
        }
    }

    fun sellBook(title: String, author: String, copies: Int) {
        if (copies <= 0) return

        getBookBy(title, author)?.let {
            it.removeCopies(copies)
            inventory.removeBookIfNoMoreCopies(it)
        }
    }

    protected fun getBookBy(title: String?, author: String?): Book? {
        return inventory.find { it.title == title && it.author == author }
    }
}

File: solution/c#/week02-object-calisthenics/ExerciseBookStore/BookStore.cs
namespace Exercise2BookStore;

public class BookStore {
    private readonly BookInventory _inventory = [];

    public void AddBook(string? title, string? author, int copies) {
        if (copies <= 0)
            return;

        var book = GetBookBy(title, author);
        if (book != null) {
            book.AddCopies(copies);
        } else {
            var newBook = Book.TryCreateBook(title, author, copies);
            if (newBook != null) {
                _inventory.Add(newBook);
            }
        }
    }

    public void SellBook(string title, string author, int copies) {
        if (copies <= 0)
            return;

        var book = GetBookBy(title, author);
        if (book != null) {
            book.RemoveCopies(copies);
            _inventory.RemoveBookIfNoMoreCopies(book);
        }
    }

    protected Book? GetBookBy(string? title, string? author) {
        return _inventory.Find(b =>
            b.GetTitle() == title && b.GetAuthor() == author);
    }
}